### ISSUE TITLE
[FEATURE]: Simulate different workspaces per screen using Virtual Desktops

### ISSUE THREAD

**Issue by phrxmd** [Bismuth-Forge/bismuth#193]
### Summary

Because KWin is at its heart not a tiling window manager, it has no notion of workspaces like i3 or Sway, and you can't change the content of one screen without also changing the content of the others. 

Bismuth could provide a limited version of "simulated workspaces", i.e. give the user the experience to change what is on one screen while leaving in place what is on the other. This could be done by having (in the background) different virtual desktops - KWin would always show the same virtual desktop 1, and use other virtual desktops to store the content of other workspaces, and when you switch to a different "workspace", Bismuth would fetch what is on the virtual desktop for that "workspace" on the current screen and exchange it with what's there at the moment.  

### A clear and concise description of what you want to happen

Some "true" tiling window managers treat the different screens like different desktops. For example, in I3 I can have several workspaces, and I can assign workspaces to particular outputs. In that way, if I have two Screens 1 and 2, I can have some workspaces always on Screen 1, others on Screen 2. When I switch workspaces on Screen 1, the content on Screen 2 stays the same.

In KWin this is not directly possible. I can have several virtual desktops, but the virtual desktop always covers all screens. So when I want to switch virtual desktops to change what is shown on Screen 1, Screen 2 changes as well. 

I would like to have a way to switch the content of Screen 1, while leaving the content of Screen 2 the same.

One way this could work is to use KWin's virtual desktops. I could have a virtual desktop [VD] 1 which is shown by KWin, and several other virtual desktops VD 2, ..., n that contain other applications. In i3's teminology, VD1 contains my "output", and VD2,...,n my "workspaces". Bismuth would need to have a way to track that on Screen 1 I want to see what is on Workspace X1, and on Screen 2 what's on Workspace X2. When I want to switch the content of Screen 1 from X1 to Y1, Bismuth could move all Screen 1 windows on VD 1 to another virtual desktop (e.g. "back" to VD X), and fetch all Screen 1 windows from VD Y to workspace 1. The Screen 2 windows would remain untouched.

In that way, in my normal workflow I would never switch virtual desktops in KWin, instead I would use Bismuth's "workspace-switching facility" to fetch windows from other VDs to my main one. Because virtual desktops in KWin always have the same geometry, desktop background, panels etc., the basic user  experience would be very similar to what I get in i3. 

It would still be more limited than i3 in that what is on Screen N would always be on Screen N. I would not be able to take a workspace and move it from one screen to another - no way to have a workspace appear on the left today and on the right tomorrow. However, from my time using i3 and Sway I didn't do that much anyway, and even the basic version as outlined above would help me a lot.

**Comment by gikari** [comment]
Do I understand correctly, that what you want is separate virtual desktops per screen, like in Sway? (I used Sway, so bring the example from it). I.e. 3 desktops for screen 1 and 6 for the screen 2?

**Comment by phrxmd** [comment]
What I want is something similar to the Sway-like behaviour that I can change what is shown on screen 1, while what is shown on screen 2,...,n  stays the same. I haven't used Sway much, but in i3 you have the `workspace next_on_output` and `workspace prev_on_output` commands that do this.

It's easy to do this with workspaces, you can choose whatever workspace you want to see on which output. You want to change what is shown on screen 1 - you go to a different workspace for this output, and what's on the other outputs stays the same. But in KWin we don't have workspaces, we only have virtual desktops. Also KWin imposes some restrictions: the number of virtual desktops is always global, and you can't switch virtual desktops just for one screen. Therefore we can't replicate the Sway behaviour 1:1, and we would have to do some tricks.

So if we want to change the content of only one screen (say screen 1), and leave the content of screen 2,3 etc. the same, we have only two options:
1. move the windows on screen 1 away - e.g. by putting them on another virtual desktop (that is not shown) - and fetch the new content for screen 1 from there (screen 2,3 etc. stay the same);
2. change to another virtual desktop that contains what we want to show on screen 1, and populate this virtual desktop on screens 2,3 etc. with the windows from the virtual desktop where we were before.

Maybe I'm going too far into the solution space already, maybe there are other ways to do it.

**Comment by gikari** [comment]
> What I want is something similar to the Sway-like behaviour that I can change what is shown on screen 1, while what is shown on screen 2,...,n stays the same. I haven't used Sway much, but in i3 you have the `workspace next_on_output` and `workspace prev_on_output` commands that do this.

So, do you want to replicate `workspace next_on_output` command for i3?

**Comment by phrxmd** [comment]
Something like that, yes.
The UX should be like "change virtual desktops, but only on one screen".

**Comment by gikari** [comment]
So, I will rephrase your proposal to make sure I fully understand.

You want a feature called "workspaces": each display/screen has a surface X by Y pixels, that could have some windows - it is called "workspace". Workspaces are dynamic: you could assign each to the particular screen. For example, screen 1 has three workspaces, which can have any number of windows, and screen 2 has two workspaces, that can also have some windows.

This functionality contradicts (i.e. its mutually exclusive to) the existing functionality of KWin - virtual desktops. Virtual Desktop is when you have one surface, that is not always rectangular, that spans all the monitors. This surface is called virtual desktop. You can have multiple virtual desktops, between which you can switch.

You want to have workspaces, that are somehow inside the virtual desktops. In that case, let me call it "window packs". You want to have a window packs that are independent of virtual desktops, so that you can manipulate them and throw around.

However, what you want could also be implemented natively by KWin, by implementing the "workspaces" concept as an alternative to the virtual desktops. In that case, that would be the thing people often call "Different Virtual Desktops for different monitors".

If what I just wrote is a correct understanding of what is happening and what you want, then I have the following thoughts:

This is very similar to the feature "window tab grouping" in a sense, that this feature is better implemented upstream. This is because introducing this concept will create a new entity, that should be somehow be present in 3 different places: System Settings, Plasma Applet and in window manager itself. Doing all of these requires a proper UI/UX design and in general needs to be more close to the existing Plasma facilities (pagers, kcms, etc.). This requires a lot of work and is out of scope for Bismuth, because it could be useful for people with floating windows too.

TLDR: this could be implemented, but not in Bismuth, but in KWin natively.

**Comment by phrxmd** [comment]
I think you've made things more complicated than they are.

Let's stay in KWin terminology. Let's say I have 2 screens and 3 virtual desktops.

I want to be able to switch all the windows on screen 1 with the windows from a different virtual desktop on screen 1, while leaving the windows on screen 2 in place.

For this we don't need a different entity, we just need to move windows between virtual desktops selectively.

**Comment by gikari** [comment]
> I want to be able to switch all the windows on screen 1 with the windows from a different virtual desktop on screen 1, while leaving the windows on screen 2 in place.

This sounds like a use case for which virtual desktops were invented for, except that they are spanning over multiple monitors. Without this exception there is a feature called workspaces on Sway/i3.

I understand this use case, but once again this is better implemented on the KWin level with mutual exclusivity for Virtual Desktops and proper design. There is zero sense in workspaces, if you have virtual desktops, and vice versa, so implementing them on Bismuth level is unnecessary complicated (like really, this is reinventing the wheel - Bismuth basically needs to emulate a workspace/virtual desktop - save all the properties of the windows to be able to move them between the workspaces).

Like, I'm really all hands on the workspaces feature, but it's not worth it to implement it here. KWin sources is much better place, so you better ask it there (https://bugs.kde.org)

**Comment by phrxmd** [comment]
Yes, I know that the cleanest solution would be to have workspaces per output in KWin. Doing that in KWin is major work, adding logic for a new entity besides Virtual Desktops and Activities, with the UI to manage it. Seeing how overloaded the KWin developers are, I don't see that happening soon. [There's been a feature request since 2005](https://bugs.kde.org/show_bug.cgi?id=107302).

Hence the idea of faking it by showing the same virtual desktop, but switching the windows on the active screen with those from a different virtual desktop :) 

Let's see if I can come up with a proof of concept, it might take a while, but it's a good motivation to finally learn KWin scripting. To me it seems it would fit well in the context of a tiling script, and it has some tiling implications (e.g. switching tiling layouts together with the window). But this will be easier to see when there is code.

**Comment by windowsrefund** [comment]
Chiming in here as someone who just found and started taking a look at this extension. What's needed is the ability to assign workspaces (virtual desktops, whatever KDE calls it) on a per display basis. 

In i3, this is extremely easy to do
```
# set up some variables
set $monitor-left HDMI-A-0
set $monitor-right HDMI-A-1
set $ws1 1
set $ws2 2
set $ws3 3
set $ws4 4
set $ws5 5
set $ws6 6

# assign workspaces to desired displays
workspace $ws1 output $monitor-left
workspace $ws2 output $monitor-right
workspace $ws3 output $monitor-left
workspace $ws4 output $monitor-right
workspace $ws5 output $monitor-left
workspace $ws6 output $monitor-right
```
